### ISSUE TITLE
Grafana v5 URL's contain randome letters or numbers (no more pretty URLs)

### ISSUE THREAD

**Issue by lukaszgryglicki** [grafana/grafana#11013]
Please include this information:
- What Grafana version are you using? **Version 5.0.0-beta1 (commit: 8c027ea)**
- What datasource are you using? **Influx**
- What OS are you running grafana on? **Ubuntu 17.04**
- What did you do? **Grafana v5 dashboard URLs contain random letters (upper & lower case), dashboards imported from v4 JSONs conatin number zero-padded for example 00000028 etc.** 
- What was the expected result? **Pretty URLs just like in Grafana v4**
- What happened instead? **URLs with random letters or numbers**

Example URLs:
- https://prometheus.devstats.cncf.io/d/UeP_tSqkz/commits-repository-groups?orgId=1
- https://prometheus.devstats.cncf.io/d/000000001/dashboards?refresh=15m&orgId=1
- https://coredns.devstats.cncf.io/d/000000018/prs-approval?orgId=1
- https://k8s.devstats.cncf.io/d/hnphTo3kk/pr-workload?orgId=1

All of them had a form "https://projectname.devstats.cncf.io/dashboard-name?orgId=1 in Grafana v4.

**Comment by lukaszgryglicki** [comment]
OK, thanks for an info.
Would be great if you can make that random string optional.

**Comment by lukaszgryglicki** [comment]
Is it possible to change this UID manually in JSON
To somethig very short, like "c".
Does it have to be uinque for all dashboards?

I've tried to do that, and if I change uid manually and try to reimport the dashboard - it is no longer possible.

**Comment by marefr** [comment]
@lukaszgryglicki yes, you can change the UID through the HTTP API if you do an update of dashboard using existing ID and a new UID. The UID is a unique identifier for all dashboards (and folders). If you import a dashboard that has a UID set, it will be imported with this UID.

But please make sure to update to the latest beta release since we've had some issues with this.

**Comment by lukaszgryglicki** [comment]
OK so I'll wait a bit, because I've also had issues (when I've changed any uid it was no longer importable), probably too old version and I don't wasnt to update all 36 instances now.
I'll do another round tomorow or some other day, thanks.

What do you mena by "HTTP API"?
I was rather thinking about chnaging uid using plain text editor and just trying to import changed JSON file.
What should I do instead?

**Comment by torkelo** [comment]
are you using the new provision system? then you just need to update the uid in the json file.

**Comment by lukaszgryglicki** [comment]
Not sure what do you mean:
Grafana version is: **Version 5.0.0-beta1 (commit: 8c027ea)**
Anyway, seems like there were some issues with this, so I will just update Grafanas tomorrow or maybe today (time permitting) and then I'll check again.
Thanks for informations.

**Comment by marefr** [comment]
> What do you mena by "HTTP API"?
I was rather thinking about chnaging uid using plain text editor and just trying to import changed JSON file.
What should I do instead?

You can absolutely do that, but use latest beta release as I wrote earlier. 

I just took for granted that you had some kind of automatic update of dashboards/dashboard provisioning since you have a lot of instances and dashboards for the cncf devstats sites :smile:

**Comment by lukaszgryglicki** [comment]
OK, so I'll try newest beta. Thanks.

**Comment by dankohn** [comment]
Do you have a recommendation on best practices for automatic updates? We
have 32 instances and growing.

--
Dan Kohn <dan@linuxfoundation.org>
Executive Director, Cloud Native Computing Foundation https://www.cncf.io
+1-415-233-1000 https://www.dankohn.com

On Mon, Feb 26, 2018 at 4:38 AM, Marcus Efraimsson <notifications@github.com

**Comment by lukaszgryglicki** [comment]
I was thinking about this many times, I have a very detailes file https://github.com/cncf/devstats/blob/master/ADDING_NEW_PROJECT.md and I'm following it every time.
The reason we don't have automatic provisioning of new dashboards is that each case is rather different that previous one, and I don't see a good way of automating this.
Currently I'm copying dashboards from one projects and do quick search & replaces in Vim/ed.
When moving those dashboards from test to prod I'm just taking full grafana.db file, copying it via sftp and then login to new prod instance to change IPs, passwords, user names etc.
It is faster than reimporting all test dashboards and works OK.
But I'm **very** open to any ideas about automating this.

**Comment by marefr** [comment]
Please check out new [provisioning feature](http://docs.grafana.org/guides/whats-new-in-v5/#provisioning-from-configuration) in Grafana v5.0. With that you can provision datasources and dashboards from the file system. In the future there will possibly be more options available than just file system.

**Comment by lukaszgryglicki** [comment]
Will do.

**Comment by lukaszgryglicki** [comment]
Usually all I need in a new dashboar is some way of templating standard project name used in various contexts. For example let's if project name is "CoreDNS".
- We will use its full name everywhere when it is visible on the UI, like descriptions etc.
- We will use "coredns" as a label/tag and in few other areas.
- We will use "DS_COREDNS" as data source name.

Usually when I copy dashboards from one project to another, I'm then staring vim on all copied dashboards and apply:
- `%s/"oldproj"/"newproj"/g|%s/DS_OLDPROJ/DS_NEWPROJ/g|%s/OldProj/NewProj/g|w|next`

But what I really need is some kind of templating.
Actually I can just use Grafana variables, create 3 variables for project name like `[[full_project_name]]`, `[[lowercase_project_name]]` and `[[datasource_name]]` and the just use them.
I think I don't need data source provisioning, because there is always only one data source in our case - InlfuxDB with already generated series.

So I'm not sure if I can use provisioning in our case, I think standard variables may be simpler but I'm not sure if I can use them everywhere and for everything that is in dashboard's JSON - like datasource name that comes from template variable etc.

**Comment by dankohn** [comment]
The goal is GitOps. No manual configuration on the servers needed.

We eventually want to get a CD system where committing on the production
branch automatically causes the production servers to upgrade.

--
Dan Kohn <dan@linuxfoundation.org>
Executive Director, Cloud Native Computing Foundation https://www.cncf.io
+1-415-233-1000 https://www.dankohn.com

**Comment by lukaszgryglicki** [comment]
@torkelo can you please answer one question?
- Can I use template variable for labe name?
![screen shot 2018-02-26 at 12 54 37](https://user-images.githubusercontent.com/2469783/36669175-4d03852c-1af4-11e8-88cd-f136ed3a2ce4.png)
![screen shot 2018-02-26 at 12 55 07](https://user-images.githubusercontent.com/2469783/36669176-4d21fb06-1af4-11e8-80e1-02764c2419b9.png)

I have a very simple variable `label_name` - I'm using it everywhere on this dashboard.
It works almost everywhere - but not for dashboard label name
Tried `[[label_name]]` and `$label_name` notations?

Is it supported at all?

**Comment by torkelo** [comment]
no, you cannot use a variable in a dashboard tags, how would that work? Dashboard tags are saved indexed (in a dashboard_tag table). Not sure what your trying to do using a variable in the dashboard tag.

**Comment by lukaszgryglicki** [comment]
OK.
I wanted to use variable from Influx to automatically tag dashboards.
That way I would have different tags when selecting different Influx database.
Nevermind, I can live without this, I was just curious if I can do it.

**Comment by lukaszgryglicki** [comment]
OK, I've updated Grafanas to `Version 5.0.0-beta5 (commit: 7fb34f7)` but when I change `uid` locally in a test editor and import dashboards - it still has old `uid` in URL and when I choose `View JSON` I see old `uid` there.
the only difference is that I can import dashboard now, but it ignores `uid` field.

**Comment by torkelo** [comment]
If there is a db id field in json that will be used and then the existing uid wil be used. 

Remove the existing dashboard and do the db (int ) id from json before saving

**Comment by torkelo** [comment]
Just removing the id field might also work, if you have overwrite flag set to true , will find existing dashboard based on name and overwrite that. But not 100% sure

**Comment by lukaszgryglicki** [comment]
Hmm - dashboard has "id": null, there are multiple other "id" fields - but they're panels ID's. All of them different 1,2,3,4.
Tried changing uid to just "c" and also tried to manually update "version" to be higher that the newest one in dashboard history.
Seems like uid is just ignored now, when changed manually in JSON.

**Comment by torkelo** [comment]
are you using the save dashboard api or the UI import api?

**Comment by lukaszgryglicki** [comment]
First I'm exporting the dashboard:
"share dashboard" -> "export" -> "save to file" -> saving it as filename.json
Then I'm editing that JSON - changing `uid` field.
Then I'm clicking the "+" icon -> "import" -> "upload .json file" -> choosing edited JSON -> selecting InfluxDB datasource and finally click "Import (overwrite)" (it says that dashboard with that name already exists which is OK"
Then I still see old uid in URL and when I click "setting" -> "view json" I see old uid.

**Comment by torkelo** [comment]
yea, overwriting an existing dashboard will resuse that dashboards ids & uid. Will have to remove it first if you want to change uid

**Comment by lukaszgryglicki** [comment]
Ok thanks, worked, changed uid to "c" here: https://k8s.cncftest.io/d/c/activity-repository-groups?orgId=1

**Comment by lukaszgryglicki** [comment]
Even better, I've deleted and reimported another one with THE SAME uid="c": https://k8s.cncftest.io/d/c/commits-repository-groups?orgId=1
And it worked too.

**Comment by lukaszgryglicki** [comment]
The question is:
- Can I use the same uid everwhere?
- Is it working because it is perfectly supported, or is it worked because it is bug?

Name`uid` sounds like something that should be unique ...
@torkelo ?

**Comment by marefr** [comment]
@lukaszgryglicki if you navigate the links you'll notice that they will take you to the same dashboard. So you can use one uid for one dashboard per grafana installation.

**Comment by lukaszgryglicki** [comment]
Ohh yes indeed.
So OK all is finally clear.
Uid must be unique, but at least I can change it to something more human-friendly.
Thanks.
Sorry for bothering you.

**Comment by marefr** [comment]
If you import a dashboard having an uid of existing dashboard and a title that not exists it will update the existing dashboard based on uid. 

If you import a dashboard having a uid that not exists and a title that exists it will update the existing dashboard based on title and it's existing uid will not change. 

Are there any other use case that you're referring to regarding needed warning/error?

**Comment by lukaszgryglicki** [comment]
I was importing a dashboard that having an existing uid "c" and existing title "commits repo groups".
Existing dashboard was "activity repo group" with uid "c"
After the import "activity repo groups" is gone and I only have "commits repo groups" with uid "c"

In this case it deleted existing dashboard (with the same uid) without warning.

**Comment by marefr** [comment]
That sounds like a bug. Going to verify this. Thanks

**Comment by marefr** [comment]
@lukaszgryglicki I may have been too fast there. What you're referring to is what I wrote earlier:
> If you import a dashboard having an uid of existing dashboard and a title that not exists it will update the existing dashboard based on uid.

That's by design. In this case it will overwrite the existing dashboard with the updated json model. Do you mean that you didn't get a overwrite warning on the import page for this case?

**Comment by lukaszgryglicki** [comment]
I was meaning that both uid and title were existing.
You were writing about:
```
If you import a dashboard having an uid of existing dashboard and a title that not exists it will update the existing dashboard based on uid.
```
But probably you're righ, if two uids are the same it overwrites.
Not a problem for me, I've just changed uids to be unique and imported again and now all is super fine:
https://k8s.cncftest.io - all UIDs are just integers and all works fine.

I was just wondering if importing dashboard with the same uid overwrites some other dashboard without any info or question - something like "dashboard with uid "c" already exists - do you want to overwrite it?"

Even more - it may have asked to overwrite, but I knew that dashboard with that **name** exist and I've imported - it overwritten the other dashboard instead.

This is the case that both name and uid are already existing - I expected that it will overwrite the dashbord that I've just modified - it overwritten the other - using uid.
Just a matter of priority, but this **may** be confusing for the user, so you can consider some message that will say what exactly will happen if you import that file?

As I said - I'm OK with this and already know how it works, but new user can possibly be suprised (as I was).

**Comment by marefr** [comment]
Yeah that's a good point. I think that we'll implement the same overwrite logic as when saving a dashboard to a title that's already exists - then a config dialog pops up asking the user. But will probably not be implemented until v5.1.

Thanks for feedback!

**Comment by lukaszgryglicki** [comment]
No problem.
